[RFC] Support gettimeofday() and settimeofday()

askawu opened the pull request:
By sntp client (#1374), we can get the current time from the network.  And I decided to give it a try to enable these two functions, so the application can set and get the current time. The idea is to make libc access system time through RTC driver API.

For RTC driver API, a new function "rtc_set_time()" is added to set the current time without using rtc_set_config() which includes alarm setting and callbacks.

To prove the idea, frdm k64f RTC driver is added based on mcux sdk and some changes are made to dts and yaml files. A sample app is also added to test gettimeofday() and settimeofday().

I don't know if this makes sense to everyone. So I create this RFC for the comments.

TODO:
- Need to add more description to RTC driver API documentation. the value passed to and get from driver api is not clearly defined.
- The rtc_set_time() implemenation of rtc_qmsi.c

Commit:
dts: common: Add RTC yaml definition

Add yaml definition in order to extract configurations from rtc node.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

Commit:
drivers: rtc: Add frdm k64f rtc driver

This patch implements frdm k64f rtc driver based on mcux sdk. Yaml and
fixup files are added to support device tree. dts files are modified for
configurations.

convert_datetime_to_seconds() and convert_seconds_to_datetime() are
copies of mcux private functions to prevent modifying mcux sdk. So
it won't be a problem when upgrading mcux sdk to upstream.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

Commit:
samples: epoch_time: Add epoch time sample app

Test gettimeofday() and settimeofday().

Signed-off-by: Aska Wu <aska.wu@linaro.org>

pfalcon commented:
As a quick comment, this is really useful! And we had this being asked for on the mailing list at least couple of times. Didn't look at the specific patches yet, will do as time permits. But definitely worth continuing IMHO, I'd like to see this feature merged.

tbursztyka commented on include/rtc.h:
```diff
@@ -108,6 +111,14 @@ static inline int rtc_set_config(struct device *dev,
+
+	return api->set_time(dev, seconds);
+}
+
```
all the above RTC API modification should go in a separate patch first, then the implementation on the driver you want, and finally this patch for newlib. It's much clearer in git history and the perhaps future changes.

lpereira replied:
Even then, it shouldn't take only seconds since epoch; it should take something with higher resolution. Maybe seconds+microseconds like in `struct timeval`.  If at all possible, it should take a u64_t for seconds: 2038 is just around the corner, and for an embedded operating system, who knows for how many years 1.10 will be used for.

askawu replied:
@lpereira, Agree. u64_t should be used. However, I found the size of time_t in new lib is still 32 bits and new lib changes to use 64 bits recently by this patch: https://sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=commit;h=4de8754bac676fc7f29082ccf8281b1c73ce629e. We probably need to apply this too.

lpereira replied:
Yes, we'll need to update newlib eventually.

mnkp replied:
Variables storing RTC values are often defined as signed to allow 'going back in time' and to facilitate signed arithmetic. Should we do the same?

tbursztyka reviewed (changes requested):
Nice addition. Just one comment on the new lib patch that would need to be split in 3.

andyross commented on lib/libc/newlib/libc-hooks.c:
```diff
@@ -157,3 +162,51 @@ void *_sbrk(int count)
+	}
+
+	if (tv) {
+		tv->tv_sec = rtc_read(rtc_dev);
```
This works, but note that on most OSes this function is used for hot path benchmarking and stuff: it tends to be implemented with a layer above the RTC hardware (e.g. on x86 linux it's done with kernel-maintained calibration on top of the TSC and isn't a system call at all).

It's not unlikely that you'll find your users complaining about the slowness of this function more than thanking you for the implementation.

askawu replied:
@andyross, Thanks. I got the idea. The resolution is not enough. And access RTC hardware directly might not be a good idea because gettimeofday() might be called very frequently.

lpereira replied:
Reading the RTC hardware directly is fine for a first implementation, or if you need this in a pinch.

A better thing to do, IMHO, would Zephyr count the time itself to a 64-bit integer, and have these functions query those numbers.  This would work even on devices whose RTC registers a 32-bit wide.  Also, with the memory protection and system call thing becoming reality, the lesser we have direct hardware access for high level concepts such as time, the better.

andyross commented on lib/libc/newlib/libc-hooks.c:
```diff
@@ -157,3 +162,51 @@ void *_sbrk(int count)
+
+	if (tv) {
+		tv->tv_sec = rtc_read(rtc_dev);
+		tv->tv_usec = 0;
```
And this related bit is **really** going to annoy people.  The whole purpose to gettimeofday() in Unix (vs. the classic time() system call) is that it supports very fine time units.  Bailing and returning an integer count of seconds is not going to meet the expectations of lots and lots of real world code.

andyross reviewed (changes requested):
Patch needs splitting as Tomasz mentioned.

Also while this meets the technical definition of gettimeofday() it really doesn't do what people are going to expect when porting code that uses it.

I think a much cleaner scheme would be to work up an "augmented clock" API in the kernel which used a hardware cycle timer (c.f. k_cycle_get_32()) to provide the fine resolution on top of a RTC-maintained count of seconds.  Then you could implement this API on top of that.

dbkinder commented on samples/system_time/README.rst:
```diff
@@ -0,0 +1,31 @@
+.. _System Time Sample App:
+
+System Time Sample App
+###########
```
the heading underline must be at least as long as the title, add some more `#`

dbkinder commented on samples/system_time/README.rst:
```diff
@@ -0,0 +1,31 @@
+.. _System Time Sample App:
```
Since this sample isn't within an existing samples folder (you're creating a new samples/system_time folder), you'll also need to create the index infrastructure and linking for the documentation.  I can do that later if you'd like, so to get around the warning saying this README.rst isn't included in any TOC tree, add a line to the top of the file:

     :orphan:

dbkinder commented on samples/system_time/README.rst:
```diff
@@ -0,0 +1,31 @@
+
+Overview
+********
+This sample demostrates that gettimeofday() and settimeofday() works on frdm
```
will this sample only work on the frdmk64f board?

dbkinder commented on samples/system_time/README.rst:
```diff
@@ -0,0 +1,31 @@
+Building and Running
+********************
+
+This sample sets the system time and output system time every 5 seconds to the
```
How about:

     This sample initially sets the system time using settimeofday() to 
     Sep 14, 2017 20:02:02 GMT (150529000) and then calls gettimeofday()
     every 5 seconds, displaying the seconds value to the console.

     You can build and execute this sample on a frdm k64f board with:

dbkinder commented on samples/system_time/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+sample:
+  name: system time sample app
+  description: gettimeofday() and settimeofday works on zephyr project
```
should be consistent with the `()` (either use on both function names, or neither)

dbkinder commented on samples/system_time/src/main.c:
```diff
@@ -0,0 +1,26 @@
+
+	tv.tv_sec = 1505290000;
+	rv = settimeofday(&tv, &tz);
+	printk("settimeofday: %d\n", rv);
```
maybe:

     printk("settimeofday: return status=%d\n", rv);

dbkinder commented on samples/system_time/src/main.c:
```diff
@@ -0,0 +1,26 @@
+	while (1) {
+		k_sleep(5 * 1000);
+		rv = gettimeofday(&tv, &tz);
+		printk("gettimeofday: %d, sec: %ld\n", rv, tv.tv_sec);
```
similarly:

     printk("gettimeofday: return status=%d, sec=%ld\n", rv, tv.tv_sec);

dbkinder commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -641,7 +641,12 @@ def generate_include_file(defs, args):
         sys.stdout.write("\n")
 
-        maxlength = max(len(s + '#define ') for s in defs[node].keys())
+        max_dict_key = lambda d: max(len(k) for k in d.keys())
```
Hmmm.. is this part of your patch?

andrewboie replied:
@dbkinder can you please clarify your question?

per the commit message: "If there's any key in the alias which length is larger than other keys
in the node, the include file will be incorrect, there will be no tab
between the key and value.

We need to take into account the max length of alias keys."

dbkinder replied:
I wasn't sure that the changes to scripts/dts/extract_dts_includes.py were actualy part of PR #1502 for get/settimeofday

andrewboie replied:
When reviewing PRs it helps to review patches individually instead of the entire diff of the patch series. This change was documented in the commit message.

dbkinder replied:
OK, but still unclear to me why there's a patch to extract_dts_includes.py mixed in with patches to a PR titled "Support gettimeofday() and settimeofday() ".  Shouldn't the patch to extract_dts_includes.py be in a separate PR?

andrewboie replied:
Have a look at the patch series in order.
First patch fixes issues in DTS script.
Second patch introduces DTS for RTC driver
Third patch implements the RTC driver.
Seems to me that the script change is a dependency for the RTC driver DTS.

dbkinder replied:
Ah. I understand the flow now.  We've had some problems where extraneous commits were showing up in PRs so I thought this might be another. Thanks.

askawu replied:
@dbkinder, I encounter the issue that the generated_dts_board.h is incorrect when I modify following changes to dts file.

```
 +&rtc {
 +       status = "ok";
 +       interrupts = <46 0>;
 +       interrupt-names = "alarm";
 +};
```
 
generated_dts_board.h:
```
#define NXP_K64F_RTC_4003D000_IRQ_ALARM_PRIORITYNXP_K64F_RTC_4003D000_IRQ_0_PRIORITY
```
You can see there's no tap between the macro and the definition. 385b340 fixes it and it's necessary to this PR.

dbkinder reviewed (changes requested):
suggested edits

andrewboie commented on samples/system_time/Makefile:
```diff
@@ -0,0 +1,4 @@
+BOARD ?= frdm_k64f
```
"Test gettimeofday() and settimeofday()."

Since this is a build_only sample and not a proper test, we really aren't testing anything.
Can you rewrite this as a proper test case which can be tested in the QEMU emulators, or at least on some hardware platform in our test case framework? Surely there must be some way we can programmatically prove that this mechanism works.

dbkinder replied:
Perhaps settimeofday, wait n seconds, and see if gettimeofday is n seconds later, where n isn't a value too large to hold up testing -- maybe 5-10 seconds?

dbkinder commented:
scripts/dts/extract_dts_includes.py is also included in PR #1502 (where I added my comment)?

Commit:
driver: rtc: Add rtc_api_get_ticks_per_sec()

The clock frequency of RTC might be configurable. The application needs
to know the frequency in order to convert the ticks to desired time
format, e.g. seconds or microseconds.

Add rtc_api_get_ticks_per_sec() in rtc driver api to provide RTC clock
frequency.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

Commit:
driver: rtc: Add rtc_api_set_time()

Provide a method in rtc driver api to set the RTC init value
independently without considering other information like alarm and
callback setting.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

Commit:
newlib: Implement gettimeofday() and settimeofday()

Implement the two functions based on epoch time.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

Commit:
time: Add epoch time implementation

This patch provides an interface to manage the epoch time based on RTC
and the hardware clock.

With the help of RTC, we can keep the epoch time even when the system
is off. When the system is up and running, we can provide a fine
resolution time unit, i.e. microseconds, based on hardware clock (though
the resolution is still depends on the clock frequency of hardware
clock).

Initially, it will read the current epoch time from RTC and keep current
hardware clock cycles as the baseline. When the epoch_time_get() is
called, the elapsed time will be calculated based on the difference of
current hardware clock cycles and the baseline. If RTC is not available,
it will assume current time is 00:00:00 UTC on 1 January 1970.

A new option "CONFIG_EPOCH_TIME_RTC_WRITE_THROUGH" is introduced. By
enabling this option, if current time is changed by epoch_time_set(),
the change will also be applied to RTC. So the time will be on the track
even if system is off.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

Commit:
kernel: add k_cycle_get() API

This function is used to get the 64-bit current time (in cycles).

For each timer driver, the default implementation of _timer_get_cycle()
are also added. The implementation just returns 0 to prevent build error
for the time being.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

Commit:
cortex_m_systick:: add _timer_cycle_get()

It returns the 64-bit accumulated clock count since system starts.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

askawu commented:
Update:

- The clock frequency of RTC might be configurable, e.g. 1 Hz or 32768 Hz. It's different between different hardware design. In 23c7080, this information is provided by rtc_api_get_ticks_per_second(), so the caller can convert the RTC ticks to the desired time format, such as second or microsecond. 

- We need to use the hardware cycle to provide the microsecond-precision measurement. However, k_cycle_get_32() only works for the short elapsed time. For example, it will overflow every 35 seconds on the 120MHz cortex m device. 8d20369 adds k_cycle_get() get a 64-bit accumulated hardware cycles. To prevent build error, the default implementation of _timer_cycle_get() is given for all platforms but it simply returns 0 for the time being. e8c18c9 implements the _timer_cycle_get() for cortex_m3.

- 3d19d81 implements the interfaces to manage the epoch time. epoch time is calculated based on RTC as the initial time and hardware cycles as the elapsed time. New epoch time can be set by epoch_time_set() and the time of RTC will also be updated if CONFIG_EPOCH_TIME_RTC_WRITE_THROUGH is enabled. Since the implementation is based on both kernel API and RTC driver API, I  put the source code under subsys/time folder. I'm not sure if it's the right place, please let me know if any suggestion.

mnkp commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -749,14 +749,15 @@ u32_t _timer_cycle_get_32(void)
 	} while (cac != clock_accumulated_count);
 
-	return cac + count;
+	return (u32_t)(cac + count);
```
Since we're going to perform 64-bit arithmetic anyway shouldn't we implement this function as
```
return (u32_t)_timer_cycle_get();
```

mnkp commented on include/rtc.h:
```diff
@@ -137,6 +141,25 @@ static inline int rtc_get_pending_int(struct device *dev)
+ *
+ * @retval How many RTC ticks in one second.
+ */
+static inline int rtc_get_ticks_per_sec(struct device *dev)
```
If we add this function I think we need to deprecate the following macro
```
 #define RTC_ALARM_SECOND (32768 / (1UL << RTC_DIVIDER))
```
since its enforcing ticks per second value to 32768.

askawu replied:
Agree. this one should be deprecated.

mnkp commented on include/rtc.h:
```diff
@@ -137,6 +141,25 @@ static inline int rtc_get_pending_int(struct device *dev)
+ *
+ * @retval How many RTC ticks in one second.
+ */
+static inline int rtc_get_ticks_per_sec(struct device *dev)
```
To match `rtc_api_get_ticks_per_sec` this function should return u32_t.

askawu replied:
Thanks. I will correct it.

kzyapkov commented:
I also played with the idea of enabling timekeeping and calendar functions with newlib, since the STM32 RTC timer already does calendar time and requires gmtime and mktime. Too bad this is being abandoned.
